### PR TITLE
Add a no-metadata JAR variant

### PR DESCRIPTION
**Pull request by tadfisher**
Ref #1642 

This adds a JAR target without included metadata. Third parties using this
variant may supply their own metadata via a custom MetadataLoader.

---
This change is [Reviewable](https://reviewable.io/reviews/googlei18n/libphonenumber/1643)

### CODE REVIEW

**Comment by keghani** [comment]
Thanks for the pull request @tadfisher!

Will this upload the no-metadata jar to Maven Central? Or will it be a local artifact only?
If the latter, why add this complexity instead of working on the regular jar? Some local work will need to happen anyway, to copy the data files over to the APK, and isn't it better to get those from the jar?

E.g. stripping data from the jar (quick hack, should probably use jar commands and probably doable wtihout the tempdir):

```shell
mkdir -p /tmp/tempdir
unzip -oq libphonenumber-x.y.z.jar -x com/google/i18n/phonenumbers/data/* -d /tmp/tempdir
pushd /tmp/tempdir > /dev/null
zip -oqr libphonenumber-no-metadata.x.y.z.jar .
rm -rf /tmp/tempdir
```

**Comment by tadfisher** [comment]
In a third-party wrapper, it would be much better to depend directly on a Maven artifact rather than repackaging the jar.

- Consumers may freely replace the wrapper with the metadata-included dependency in unit tests or cross-platform build variants.
- Consumers may freely override the transitive dependency with a later version, not having to wait for the wrapper to catch up.
- Having a metadata-less base artifact reduces parallel work if someone wanted to change the metadata-loading facility for another use case, for example loading the rules from a database or the network.

Yes, that still means copying the metadata from the main artifact when building the wrapper, but that is a compile-only dependency for the wrapper, not a runtime dependency for the consumer.

**Comment by tadfisher** [comment]
See also https://github.com/JodaOrg/joda-time/issues/220 and https://github.com/ThreeTen/threetenbp/issues/29 for similar use cases.

**Comment by keghani** [comment]
Thanks for the details @tadfisher! I'll discuss with the team and get back to you this week.

File: java/libphonenumber/pom.xml [hunk 1]
```diff
@@ -42,11 +42,32 @@
...
         <artifactId>maven-jar-plugin</artifactId>
         <version>3.0.2</version>
-        <configuration>
-          <archive>
-            <manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>
-          </archive>
-        </configuration>
+        <executions>
+          <execution>
+            <id>default-jar</id>
```
Comment (keghani) [inline java/libphonenumber/pom.xml]: Should this have a `phase` and `goals`, and a `classifier` in the `configuration`, to be consistent with the `execution` below?

Comment (tadfisher) [inline java/libphonenumber/pom.xml]: I followed the [maven-jar-plugin example](https://maven.apache.org/plugins/maven-jar-plugin/examples/attached-jar.html); I don't believe these are necessary to add for the default jar configuration, but are required for adding an attached classified jar.

**Comment by tadfisher** [comment]
@keghani Regarding the maven publishing question: I ran `mvn release:prepare -DdryRun=true`, and it built and had me sign the no-metadata jar as well as the standard jar. So I think you're good there. I don't see any artifact-specific configuration for the deploy or release plugin.

**Comment by keghani** [comment]
Thanks! I patched this and ran some tests with and without the changes; the main libphonenumber jar didn't change as a result of this, and diffing that against no-metadata showed me data files that were only present in the main libphoneumber jar.

Thanks for verifying with the dry run! I actually didn't know that flag :)

Squashing!

**Pull request merged**